server: thin out the main start path

tbg opened the pull request:
It has always bothered me how much was going on there in no particular
order. This commit takes the myriad of workers and one-off operations
that go into starting a Server and adds them to a method at the end.

This has the effect that `(*Server).Start` is now more clearly occupied
only with setting up the listeners, bootstrapping, and starting the Node
and SQL Server.

Release note: None

tbg committed:
server: add initServer.NeedsInit

This will come in handy in follow-up commits. This commit also
introduces a first (inconsequential) use.

Release note: None

tbg committed:
server: simplify readiness handling

Explain and simplify the ReadyFn handling in `(*Server).Start`.

This commit is identical to what's in (closed PR) #48531 except for the
placement of the readiness call.

Tangentially related to #48273.

Release note: None

tbg committed:
server: thin out the main start path

It has always bothered me how much was going on there in no particular
order. This commit takes the myriad of workers and one-off operations
that go into starting a Server and adds them to a method at the end.

This has the effect that `(*Server).Start` is now more clearly occupied
only with setting up the listeners, bootstrapping, and starting the Node
and SQL Server.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48856)
<!-- Reviewable:end -->

tbg committed:
server: centralize HTTP setup in (*Server).Start

We were setting up the authenticated HTTP handlers quite late, exposing
clients to hitting them situations before they wre registered.

Release note: None

tbg committed:
server: move UI-serving code into startServeUI

We were setting up most of the HTTP endpoints outside of startServeUI,
which cluttered `(*Server).Start`. Instead, do this setup right when
we're also ready to open and serve up the listener.

Release note: None

tbg committed:
server: add comment on debug server auth

Add comment describing current status, add TODO.

Release note: None
